### ISSUE TITLE
ginmappeR

### ISSUE THREAD

**Issue by fersolesp** [Bioconductor/Contributions#3124]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/DEAL-US/ginmappeR

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
thanks for your submission.   
```
Running ‘runTests.R’
 ERROR
Running the tests in 'tests/runTests.R' failed.
Last 13 lines of output:
  1 Test Suite : 
  ginmappeR RUnit Tests - 1 test function, 1 error, 0 failures
  ERROR in /home/vincent/TEMP/RtmptjZ2HO/RLIBS_430b16096f61/ginmappeR/unitTests/test_KEGG.R: Error while sourcing  /home/vincent/TEMP/RtmptjZ2HO/RLIBS_430b16096f61/ginmappeR/unitTests/test_KEGG.R : Error in checkEquals(getKEGG2UniProt("mct:MCR_1292"), c("D5VCZ9")) : 
    Lengths (0, 1) differ (string compare on first 0)
  
  
  Test files with failing tests
  
     test_KEGG.R 
       /home/vincent/TEMP/RtmptjZ2HO/RLIBS_430b16096f61/ginmappeR/unitTests/test_KEGG.R 
  
```
do you know why this is happening?

**Comment by fersolesp** [comment]
Hi @vjcitn!

It seems that test was broken, I have already fixed it.

I am sorry for the inconvenience.

**Comment by lshep** [comment]
Your utilsFunctions.R look like they are writing data to a users working directory `getwd()`;  It is good practice to never write/create a directory without a users knowledge.  This should be changed to use a temporary directory by default but allow the user to specify the directory to write the data.

**Comment by fersolesp** [comment]
Hi @lshep!

Thanks for the suggestion. I have changed the code to use, by default, the tempdir() temporary directory provided by R and also added a changeCARDPath() function to let the user specify another desired path.

**Comment by hpages** [comment]
Hi @fersolesp,

Are you sure **plumber** is needed? It's listed in your `Suggests` field.

FWIW I can run `R CMD build` / `R CMD check` on the package without having **plumber** installed on my machine,  and after removing it from `Suggests`. The only mention of **plumber** is in your `README.md` file where you provide an obscure **plumber** command that is not documented. Also please note that this command will never be tested because it's only in `README.md`.

The reason I ask is that **plumber** has additional system-level requirements (`libsodium-dev` on Ubuntu) that we would need to install on our build machines, only for **ginmappeR** to pass `R CMD check`, even though **plumber** won't ever be used. One way to avoid this is that you put **plumber** in `Enhances` instead of `Suggests`.

Also please provide details about what that obscure command does in your `README.md`.

Thanks,
H.

**Comment by fersolesp** [comment]
Hi @hpages!

Plumber was intended to allow users to use ginmappeR functionalities through an Application Programming Interface (API), and the launch command was the one in the README. So it is not really needed.

As it is an extra feature, for the moment, I have disabled it and deleted Plumber from Suggests. Now the build is successful as it does not need the additional system-level requirements.

What worries me now is the Timeout for R CMD Check, as ginmappeR unit tests take around 15 minutes to finish, because they depend on external APIs such as NCBI, UniProt or KEGG. I have already reduced the test suite but keeps getting Timeout. I don't know if it would be possible to extend R CMD Check timeout limit.

If not, I will reduce the tests again if possible.

Thanks for your time.

**Comment by lshep** [comment]
would it make sense then to implement the long test features for some of the unit tests?  http://contributions.bioconductor.org/long-tests.html

**Comment by fersolesp** [comment]
Hi @lshep!

Thanks for the suggestion! I have relocated some of the tests to be run as "longtests". Now the package is successfully built without exceeding the time limit and without any errors or warnings.

**Comment by hpages** [comment]
Hi @fersolesp,

Thanks for making these adjustments to the package.

The package has some issues:

- One major issue is that all the examples in the man pages are wrapped in `if (interactive()) {...}` statements, which means that none of them is actually evaluated by `R CMD check`. This is against Bioconductor guidelines that require that all documented functions have examples that get evaluated by `R CMD check`. This is to ensure that the examples remain valid when checked daily by the Bioconductor build system.

- The other major issue is that access to eutils.ncbi.nlm.nih.gov seems quite unreliable. For example, at the moment I'm unable to use `getNCBIGene2KEGG()`:
    ```
    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Could not resolve host: eutils.ncbi.nlm.nih.gov

    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Timeout was reached: [eutils.ncbi.nlm.nih.gov] Resolving timed out after 10000 milliseconds

    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Could not resolve host: eutils.ncbi.nlm.nih.gov
    
    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Could not resolve host: eutils.ncbi.nlm.nih.gov
    
    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Timeout was reached: [eutils.ncbi.nlm.nih.gov] Resolving timed out after 10000 milliseconds
    
    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Could not resolve host: eutils.ncbi.nlm.nih.gov

    > getNCBIGene2KEGG('76524190')
    Error in curl::curl_fetch_memory(url, handle = handle) : 
      Send failure: Broken pipe
    ```
  Note that I get 3 different kinds of errors. Is this a known issue? Unfortunately, this might really hurt the usability of the package.

- When eutils.ncbi.nlm.nih.gov is available, it can be quite slow:
    ```
    > system.time(kegg_id <- getNCBIGene2KEGG('76524190'))
       user  system elapsed 
      0.338   0.008  10.751 
    ```
  More than 10 seconds to map 1 gene. The `getNCBIGene2KEGG()` function is not vectorized so the user who needs to map a set of genes will need to apply `getNCBIGene2KEGG()` in a loop. This will be very slow if they need to map a few hundreds or thousands of genes. Have you considered vectorizing the mapping functions? I see that most of them are memoized but that doesn't really help address the speed issue.

- The long tests have errors when I try to run them. Running them within R directly from the `longtests/` folder:
    ```
    > source("test_CARD.R")
    ...
    Error in file(filename, "r", encoding = encoding) : 
      cannot open the connection
    In addition: Warning message:
    In file(filename, "r", encoding = encoding) :
      cannot open file '../ginmappeR/R/CARDFunctions.R': No such file or directory
    
    > source("test_NCBI.R")
    Error in file(filename, "r", encoding = encoding) : 
      cannot open the connection
    In addition: Warning message:
    In file(filename, "r", encoding = encoding) :
      cannot open file '../ginmappeR/R/NCBIFunctions.R': No such file or directory
    ```
  Or running them via `R CMD check`:
    ```
    hpages@XPS15:~$ R CMD check --test-dir=longtests --no-stop-on-test-error --no-codoc --no-examples --no-manual --ignore-vignettes --check-subdirs=no ginmappeR_0.99.3.tar.gz
    * using log directory ‘/home/hpages/ginmappeR.Rcheck’
    * using R version 4.3.0 (2023-04-21)
    ...
    * checking tests in ‘longtests’ ...
      Running ‘test_CARD.R’
      Running ‘test_NCBI.R’
     ERROR
    Running the tests in ‘longtests/test_CARD.R’ failed.
    Last 13 lines of output:
      > cardPath <<- tempdir()
      > 
      > # GitHub Actions imports
      > # source('../../../R/CARDFunctions.R')
      > # source('../../../R/utilsFunctions.R')
      > 
      > # # Local execution imports
      > source('../ginmappeR/R/CARDFunctions.R')
      Error in file(filename, "r", encoding = encoding) : 
        cannot open the connection
      Calls: source -> source -> file
      In addition: Warning message:
      In file(filename, "r", encoding = encoding) :
        cannot open file '../ginmappeR/R/CARDFunctions.R': No such file or directory
      Execution halted
    Running the tests in ‘longtests/test_NCBI.R’ failed.
    Last 13 lines of output:
      > 
      > # GitHub Actions imports
      > # source('../../../R/NCBIFunctions.R')
      > # source('../../../R/UniProtFunctions.R')
      > # source('../../../R/utilsFunctions.R')
      > 
      > # # Local execution imports
      > source('../ginmappeR/R/NCBIFunctions.R')
      Error in file(filename, "r", encoding = encoding) : 
        cannot open the connection
      Calls: source -> source -> file
      In addition: Warning message:
      In file(filename, "r", encoding = encoding) :
        cannot open file '../ginmappeR/R/NCBIFunctions.R': No such file or directory
      Execution halted
    * DONE
    
    Status: 1 ERROR, 1 NOTE
    See
      ‘/home/hpages/ginmappeR.Rcheck/00check.log’ for details.
    ```

- If I try to map more than one ID at a time, I get an obscure error message:
    ```
    > getNCBINucleotide2KEGG(c('AY536519', 'NZ_CP059690'))
    Error in if (is.nan(variable)) { : the condition has length > 1
    ```
  Of course I should not do this (it seems that `getNCBINucleotide2KEGG()`, like the rest of the mapping functions, is not vectorized). However, it would be nice to improve error handling by systematically checking user input and failing with clear and useful error messages. This applies to all the functions in the package, not just `getNCBINucleotide2KEGG()`.

- The Contents appears above the title in the rendered vignette:
![Screenshot from 2023-10-16 10-44-55](https://github.com/Bioconductor/Contributions/assets/8810451/0fa3355e-dd5c-461f-8811-c250fbc9ae53)

- The References section has a lot of problems. Most BibTeX entries in your `knitcitations.bib` are broken e.g. misspelled package name (`UniProt. ws`) or wrong case (`keggrest`) or list of authors that mix people with package names etc... Note that entries for R packages can easily be generated with `citation()` e.g.
    ```
    > citation("KEGGREST")
    To cite package ‘KEGGREST’ in publications use:
    
      Tenenbaum D, Maintainer B (2023). _KEGGREST: Client-side REST access
      to the Kyoto Encyclopedia of Genes and Genomes (KEGG)_.
      doi:10.18129/B9.bioc.KEGGREST
      <https://doi.org/10.18129/B9.bioc.KEGGREST>, R package version
      1.41.4, <https://bioconductor.org/packages/KEGGREST>.
    
    A BibTeX entry for LaTeX users is

      @Manual{,
        title = {KEGGREST: Client-side REST access to the Kyoto Encyclopedia of Genes and
    Genomes (KEGG)},
        author = {Dan Tenenbaum and Bioconductor Package Maintainer},
        year = {2023},
        note = {R package version 1.41.4},
        url = {https://bioconductor.org/packages/KEGGREST},
        doi = {10.18129/B9.bioc.KEGGREST},
      }
    ```

 - Note that `LazyData: false` has no effect and is not needed. I would suggest that you remove it to avoid the `R CMD build` NOTE.

Best,
H.

**Comment by fersolesp** [comment]
Hi @hpages!

Thank you for thoroughly revising the package and highlighting the issues. This is how they have been addressed:
- I have removed `if(interactive()){}` statements in man pages. We really did not know they prevent
`R CMD check` to evaluate them.
- `LazyData: false` has been removed from `DESCRIPTION`.
- The vignette has been modified to have the `Title` before the `Contents`. I have also solved the problems
in `References` section.
- About the `longtests`. They are configured to be run with `R CMD check` and the thing was that some routes were
not correct (caused by a wrong configuration when defining the longtests). Now, they run well with the same command you employed:
`R CMD check --test-dir=longtests --no-stop-on-test-error --no-codoc --no-examples --no-manual --ignore-vignettes --check-subdirs=no ginmappeR_0.99.4.tar.gz`
- Vectorising input IDs: It has been implemented, although it really only serves to provide more convenience to the user to be able to map multiple IDs without the need to use loops. In practice, translations are not parallelised, which is intentional as it would involve simultaneous API calls, which they do not support. 
The given solution has the advantage of not blocking the code execution, now throws the warning of the error
and returns a `NULL` for the result, in this way, when mapping IDs in a vectorised manner, the user gets the warning
of non-mapped IDs and a `NULL` value for them, and not an execution-stopping error that makes the previous mappings get lost.
I tried different approaches to get the desired behaviour but most of them failed due to limitations of the R language.
That it is why it took me almost a week to came up/find/decide on this solution. The documentation has also been modified to point out the posibility of providing vectors of IDs as input and unit tests have also been updated to work correctly with these changes.
About the `memoisiong` of functions, it really is more useful than it seems, because there are lots of intermediate translations that the user does not see but they are cached.
- Error management: we use three or four packages that access public databases APIs. Ideally, each package should control the errors generated by the API it accesses, otherwise we would have to handle a wide variety and diversity of errors ourselves. I have taken advantage of the vectorisation of the functions to provide a joint and generalised solution to this problem of error handling, so that, although they are varied (and therefore informative), they are displayed in a more homogeneous and user-friendly way in the form of a `warning`.

I hope we have dealt conveniently with the issues. In any case, of course, any suggestions or improvements are welcome.

Thanks for your time.

**Comment by fersolesp** [comment]
Hi, @hpages! I hope you are doing well! I have seen that since the last changes I uploaded in October there is nothing new in the thread. Please let me know if there is anything I can help with or if you, or @lshep, can give me some feedback. Thanks in advance.

Yours sincerely, Fernando.

**Comment by lshep** [comment]
I'm very sorry for the delay; it should not be taking this long. I will check in with @hpages if he plans to revisit this review soon or reassign the reviewer if necessary

**Comment by hpages** [comment]
Hi @fersolesp  sorry for the delay. I'll take another look at the package in the next couple of days. Thanks for your patience.

**Comment by hpages** [comment]
Hi @fersolesp,

Thanks for your hard work on the package and for all the improvements.

Note that the unit tests fail for me:
```
> BiocGenerics:::testPackage("ginmappeR")
Loading required package: BiocGenerics

Attaching package: ‘BiocGenerics’

The following objects are masked from ‘package:stats’:

    IQR, mad, sd, var, xtabs

The following objects are masked from ‘package:base’:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,
    tapply, union, unique, unsplit, which.max, which.min

Loading required package: RSQLite
Testing .checkKEGGIdExists
Testing getKEGG2UniProt
Testing .getKEGG2NCBIDT
Testing getKEGG2NCBIProtein, getKEGG2NCBINucleotide, getKEGG2NCBIGene
Testing getKEGG2CARD
Updating CARD database data...
CARD database downloaded successfully!
Located at /tmp/RtmpZp1CWZ/card-data
Using a CARD database version downloaded on Fri 01-05-2024 16-31-38, please consider updating it with updateCARDDataBase() function.
Using a CARD database version downloaded on Fri 01-05-2024 16-31-38, please consider updating it with updateCARDDataBase() function.
Testing .checkUniProtIdExists
Testing getUniProtSimilarGenes
Error in checkTrue(length(getUniProtSimilarGenes("G0L217", clusterIdentity = "0.5",  : 
  Test not TRUE



RUNIT TEST PROTOCOL -- Fri Jan  5 16:32:03 2024 
*********************************************** 
Number of test functions: 1 
Number of errors: 1 
Number of failures: 0 

 
1 Test Suite : 
ginmappeR RUnit Tests - 1 test function, 1 error, 0 failures
ERROR in /home/hpages/R/R-4.4.r85764/site-library/ginmappeR/unitTests/test_UniProt.R: Error while sourcing  /home/hpages/R/R-4.4.r85764/site-library/ginmappeR/unitTests/test_UniProt.R : Error in checkTrue(length(getUniProtSimilarGenes("G0L217", clusterIdentity = "0.5",  : 
  Test not TRUE


Test files with failing tests

   test_UniProt.R 
     /home/hpages/R/R-4.4.r85764/site-library/ginmappeR/unitTests/test_UniProt.R 


Error in BiocGenerics:::testPackage("ginmappeR") : 
  unit tests failed for package ginmappeR
```
This is with **ginmappeR** 0.99.4 and R devel (2024-01-02 r85764) on an Ubuntu 23.10 laptop (see my `sesssionInfo()` below).

When I try to use the new vectorized capability, I get:
```
> getCARD2NCBIProtein(c("3003988", "3002535"))
Updating CARD database data...
CARD database downloaded successfully!
Located at /tmp/Rtmp1kLXJM/card-data
Warning: the condition has length > 1

NULL
```
Furthermore I don't see that this capability is covered by the unit tests. This would have caught the problem.

Also it would be good to show examples that demonstrate the vectorized capability of `getCARD2NCBIProtein`, `getCARD2NCBINucleotide()`, etc... in their man pages. Seeing is believing :wink:

My first attempt at using `getNCBIGene2KEGG()` failed:
```
> getNCBIGene2KEGG('76524190')
Warning: HTTP failure: 429, too many requests. Functions that contact the NCBI should not be called in parallel. If you are using a shared IP, consider registerring for an API key as described in the rate-limiting section of rentrez tutorial. NCBI message:
 {"error":"API rate limit exceeded","api-key":"71.212.16.155","count":"4","limit":"3"}


Warning: argument is of length zero

character(0)
```
After a couple more attempts it succeeded but this makes me question the reliability/usability of the package.

Other issues:

- Please remove `ginmappeR.html` from the `vignettes/` folder. The HTML vignette does not belong to the package source tree since it will be automatically rendered by `R CMD build`.

- Use of unneeded `paste()` statements in many places. For example:
    ```
    message(paste("CARD database downloaded successfully!\nLocated at ", paste(options("cardPath")[[1]], "/card-data", sep = ""), sep = ""))
    ```
  `message()` (as well as `warning()` and `stop()`) already takes care of pasting everything together so you don't need to do it. For example, the following code will produce exactly the same output as the above but it easier to read:
    ```
    message("CARD database downloaded successfully!\nLocated at ", options("cardPath")[[1]], "/card-data")
    ``` 
  Please visit the entire codebase of the package and make sure to remove these unnecessary `paste()` statements.

- `.checkIfCARDIsDownloaded()` internal helper displays the following message:
    ```
    Using a CARD database version downloaded on Fri 01-05-2024 14-12-22, please consider updating it with 
    updateCARDDataBase() function.
    ```
  Note that `14-12-22` is formatted like a date. If it's the time of the day, it should be formatted as such (e.g. `14:12:22`) to avoid confusion.

- The CARD database is cached only for the duration of the current session. Is that intended? Wouldn't it make more sense to have it cached in a more persistent way? Especially given the fact that `.checkIfCARDIsDownloaded()` takes care of reminding the user about the need to manually update the cache, which greatly mitigates the risk of working with a stale cache.

- Lots of code repetition that should be avoided. For example this exact same line of code is used in 15 different places:
    ```
    aroIndex <- read.csv(paste(options("cardPath")[[1]],'/card-data/aro_index.tsv',sep=''), sep='\t')
    ```
  Good software engineering practice is to avoid code repetition (DRY principle). Please put this code in a helper function (e.g. `.load_aro_index()`) and use that instead.
  This was just an example. Overall, the codebase in the package suffers from a lot of code repetition. Another example is the `getCARD2NCBIProtein()` and `getCARD2NCBINucleotide()` functions which are almost identical except for the last operation when the column to return to the user is extracted (`Protein.Accession` column vs `DNA.Accession` column). Please refactor to avoid the repetition.

Best,
H.

**sessionInfo():**
```
R Under development (unstable) (2024-01-02 r85764)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 23.10

Matrix products: default
BLAS:   /home/hpages/R/R-4.4.r85764/lib/libRblas.so 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: America/Los_Angeles
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] XML_3.99-0.16       rentrez_1.2.3       httr_1.4.7         
[4] KEGGREST_1.43.0     UniProt.ws_2.43.0   RSQLite_2.3.4      
[7] BiocGenerics_0.49.1 RUnit_0.4.32        ginmappeR_0.99.4   

loaded via a namespace (and not attached):
 [1] bit_4.0.5               jsonlite_1.8.8          dplyr_1.1.4            
 [4] compiler_4.4.0          crayon_1.5.2            BiocBaseUtils_1.5.0    
 [7] filelock_1.0.3          tidyselect_1.2.0        Biobase_2.63.0         
[10] blob_1.2.4              bitops_1.0-7            Biostrings_2.71.1      
[13] IRanges_2.37.0          progress_1.2.3          png_0.1-8              
[16] fastmap_1.1.1           R6_2.5.1                XVector_0.43.0         
[19] rjsoncons_1.1.0         generics_0.1.3          curl_5.2.0             
[22] GenomeInfoDb_1.39.5     tibble_3.2.1            AnnotationDbi_1.65.2   
[25] GenomeInfoDbData_1.2.11 DBI_1.2.0               pillar_1.9.0           
[28] rlang_1.1.2             utf8_1.2.4              cachem_1.0.8           
[31] bit64_4.0.5             memoise_2.0.1           cli_3.6.2              
[34] magrittr_2.0.3          zlibbioc_1.49.0         digest_0.6.33          
[37] httpcache_1.2.0         dbplyr_2.4.0            hms_1.1.3              
[40] lifecycle_1.0.4         BiocFileCache_2.11.1    prettyunits_1.2.0      
[43] S4Vectors_0.41.3        vctrs_0.6.5             glue_1.6.2             
[46] stats4_4.4.0            RCurl_1.98-1.13         fansi_1.0.6            
[49] tools_4.4.0             pkgconfig_2.0.3        
```

**Comment by fersolesp** [comment]
Hi @hpages!

Thank you for your revision. 

The execution of man pages examples in Bioconductor Package Builder R CMD check resulted in exceeding the 10 min execution limit. We already configured longtests and divided the test suite. Considering that all the functionality showed in examples is already covered in tests, would it be acceptable to wrap the examples in `\dontrun`, as suggested for package [CTexploreR](https://github.com/Bioconductor/Contributions/issues/3138#issuecomment-1761497904), as long as all functionality is tested once at some point? Alternatively, in our local tests, R CMD check seems to take ~20 minutes to run fully currently. Would it be acceptable to leave as is, since the TIMEOUT limit for accepted packages is 40 minutes? Your advice on this would be greatly appreciated.

With respect to the revision, this is how the issues have been addressed:

- The vectorizing of functions not working seems to be because of the latest update of R language: the vectorizing and memoising seems to not be working if defined, as previously done, in `.onLoad` function, so they have been relocated so they do work as expected. This made it necessary to change `memoise` cache, so instead of an R object stored in R `options()`, it is a file located in R's temporary folder: `cachem::cache_disk(paste(tempdir(),'/cache',sep='')` and more easily accesible through dedicated internal function `.getCache()`.
- Some of the tests were broken because of changes in APIs' content, but ginmappeR's functionality was still up to date. In any case, this issue has been solved and the test suite has been redone to include vectorized functions, too. We have also added vectorized examples in man pages and vignette so users can be aware of this capability.
- `ginmappeR.html` has been removed from `vignettes/` folder, as suggested.
- Displayed messages have been updated to show the date and time formatted correctly.
- The code has been thoroughly revised and refactored so it has less repeteated code and it is as maintainable and modular as possible. For example:
    - Function `.loadAROIndex()` has been created, to avoid repeating CARD loading code.
    - All unnecesary `paste()` have been deleted, as suggested. We were not aware `message()` and similar already paste everything together, so we thank you for that.
    - We have created functions `.errorMessageHandler()` y `.warningMessageHandler()` to appropriately handle errors and warnings.
- With regard to the cache of CARD database, the fact that it is cached only during the current session is partly intended: at first, we stored CARD in a folder within the user's working directory. We were told this is not a good practice and changed it to use, by default, the `tempdir()` temporary directory provided by R, that seems to be cleaned up and changed between sessions. This way has the advantage that anytime the user starts a new session, he will get the latest version without having to update it manually and considering that the compressed CARD file is about 4 MB, it is not a big deal. In any case, the user also has the option to set a custom folder for CARD to be stored through `changeCARDPath()` function. We have been looking for another solution but this one still seems to be the best one.

Regarding the problem of using `getNCBIGene2KEGG()` and obtaining a NCBI's API warning: ginmappeR makes use of three packages that access public databases free web APIs through HTTP requests that may fail. We have done our best effort to give a proper error management so that the user does not have to deal with this kind of warnings but receives a more friendly message that makes him aware of this situation and kindly ask him to try again. We have also introduced a mechanism to automatically repeat the requested mapping several times if an error arises (further explained below). For example, if we try a mapping without internet conection:
```
> getNCBIGene2KEGG('76524190')
One of the web APIs failed, please try again.
NULL
```
Or if the gene is not registered in the database:
```
> getNCBIGene2KEGG('test')
The given ID "test" is not registered in NCBI Gene database
NULL
```
Note that if an error is produced, we return `NULL`, instead of `character(0)` or `list()`, that would be returned if no mapping is possible. The reason to return `NULL` is that if a single mapping fails in a vectorized set of mappings, the execution is not aborted and the already mapped genes do not get lost. This aspect has been considered in man pages too.

Finally, the main task that has been addressed is the combination of cache, vectorization and errors and warnings management, all in the context of a package. The principal issue was that because of vectorization of functions, in case of error, `NULL` is returned and the execution of the code would not stopped, as explained before. This implies that this `NULL` value would be cached, so trying to get the mapping again would result in a `NULL` directly from cache, instead of trying to get the correct mapping again. The final result came from creating interfaces for the functions, always within R limitations. Mainly, this is the proposed solution:
```
interfaceFunction <- vectorization(retryHandler(errorCacheHandling(memoise(originalFunction))))

# For example
getCARD2NCBIProtein <- vectorization(retryHandler(errorCacheHandling(memoise(.getCARD2NCBIProtein))))
```
1. The memoising (caching) of the function is performed when the package is imported with `onLoad()` function and applies to the original function.
2. Then, an error management layer is added, so if an incorrect `NULL`, result of an error, is stored, it is deleted from the cache so next time the mapping can be executed again and avoid returning a `NULL` directly.

    First, we tried to employ `memoise`'s `drop_cache(function)(params)` function. It did not modify the cache when executed inside another function and any attempt of changing its executing environment was in vain.

    The second approach we tested was directly working with `cachem` object, used by `memoise`. This cache is a dictionary of (key, value) and easily manipulable. With `.getCache()$remove(mappingHash)` we could delete the incorrect values. The problem was that it was impossible to generate the hashes the same way `memoise` does, so we could not identify what entry of the cache to remove.

    The last approach was to leverage the `interfaceFunction <- originalFunction` architecture so when an error arises, the original function provides the interface with the list of keys in cache just before introducing the `NULL` value. In that way, we can compare the list of keys before and after caching the wrong value, get its hash and remove it from cache. This way also has the advantage of keeping correct intermediate mappings that may prove useful in other translations.

3. Because of the unstable nature of APIs, we decided to add another layer that retries the requests several times if an error is raised. This mechanism has little impact on execution time and highly improves the usability of the package.
4. Lastly, a vectorization layer is added. We employ R base `Vectorize()` function and also manage the returned result because it always returns a list, even if there is only a single element, which would break the package when functions call each other.

This estructure was the only one that resulted in the desired behaviour and with a pretty modular code. Interfaces would be created like:

```
# Wrapper function to vectorize and handle single value output
.vectorizeFunction <- function(func, vargs) {
    vectorizedFunc <- Vectorize(func, vectorize.args = vargs, USE.NAMES = FALSE, SIMPLIFY = FALSE)
    return(function(...) {
        result <- vectorizedFunc(...)
        if (length(result) == 1) {
            return(result[[1]])
        } else {
            return(result)
        }
    })
}

getCARD2NCBIProtein <- .vectorizeFunction(
    (function(f){
        return(function(cardId) {
            return(.retryHandler(f, cardId)$Protein.Accession)
        })
    })(.getCARD2NCBIAux), c('cardId'))
```

However, we had to instantiate the interface functions individually to define their default parameters and values, so that `R CMD check` would recognise these parameters and they would be the same as in the man pages. If this is not done and we use a generic function (see `.vectorizeFunction` above) that treats the parameters as `...`, then the check process will find that these `...` are different from the actual parameters in the man pages documentation.

For this reason, interfaces had to be created as shown below, making explicit the parameters received by the functions. For example:

```    
getCARD2NCBIProtein <- function(cardId) {
    # Second, we vectorize the original function
    vectorizedFunc <- Vectorize(
        (function(f){
            # First, we wrap the memoised function with a layer that retries the request if errors are 
            # raised and also handles them
            return(function(cardId) {
                return(.retryHandler(f,cardId)$Protein.Accession)
            })
        })(.getCARD2NCBIAux), vectorize.args = c('cardId'), USE.NAMES = FALSE, SIMPLIFY = FALSE)
    # Third, we return the vectorized function, expliciting the original function parameters', default
    # values and parsing the result (gets the value if the result list is of size 1)
    return((function(cardId) {
        return(.resultParser(vectorizedFunc(cardId)))
    })(cardId))
}
```
In addition to this, we had to modify the original functions so that when they call another ginmappeR function, they do not call the interface, which handles the errors and does not raise them to the calling function; instead, they call the original function. For example, if an user calls `getCARD2UniProt('3002535')`, inside of it, it calls `getNCBIProtein2UniProt`. What it is done now, is to call its original function (note the dot `.` before its name) and handle it in case of error as well: `.cacheErrorHandler(raiseError=TRUE, .getNCBIProtein2UniProt(...))`.

This is the best solution in terms of less code repetition and modularisation while having all the layers to work correctly together. It is also remarkable that the core functionality, the mappings themselves, is still intact, just layers have been added on top of it, so it is still perfectly maintainable and upgradable as it was before.

Thanks for your time,

Fernando.

**Comment by hpages** [comment]
All these changes sound good. Thanks for the hard work!

In particular, it's great that I can now do this:
```
getCARD2NCBIProtein(c("3003988", "3002535"))
```
However, surprisingly the result is a list of length 2 and not a character vector of length 2! The way vectorized functions usually behave is that if a length-1 input produces an atomic vetor of length 1, then a length-N input is expected to produce an atomic vector or length N.

For example this is what vectorized functions `sqrt()`, `toupper()`, and `as.numeric()` functions in base R do:
```
> sqrt(4:-2)
[1] 2.000000 1.732051 1.414214 1.000000 0.000000      NaN      NaN
Warning message:
In sqrt(4:-2) : NaNs produced

> toupper(c("Sponge", "Bob"))
[1] "SPONGE" "BOB"   

> as.numeric(c("5", "-0.5", "3i", "8"))
[1]  5.0 -0.5   NA  8.0
Warning message:
NAs introduced by coercion 
```
Also note that when an input value cannot be mapped, the corresponding output value is `NA` (or `NaN`). Using `NA` to "fill the gaps" allows to preserve positionality between input and output, which is a nice property that your current solution already has (via the insertion of `NULL` elements in the returned list).

Any reason why `getCARD2NCBIProtein()` and family cannot follow this widely consensual approach?

Thanks again,
H.

**Comment by fersolesp** [comment]
Hi @hpages!!

Thanks a lot for your reply!

About the fact we are returning lists instead of character vectors of the same length as the input, this is due to the fact that, when exhaustiveMapping is set to TRUE, multiple mapped IDs might be returned for each input ID, and we can only handle this through lists. We considered the possibility of flattening the lists to character vectors if the length of each element in the returned list was 1, but we believe this might lead to a change in the data type of the output (list or character vector) that would be unpredictable a priori (since it would depend on the results of the mapping itself).

Kind regards,

Fernando

**Comment by hpages** [comment]
The type of the output shouldn't change depending on the length of the input. Unfortunately, right now it does:
- if the input to `getCARD2NCBIProtein()` is a character vector of length 1, the output is a character of length 1 or a NULL
- if the input to `getCARD2NCBIProtein()` is a character vector of length 0 or > 1, the output is a list

So depending on the length of the input, the same function call can return a character vector (of length 1), a list, or a NULL! 

This behavior might be ok when using the function interactively but it will cause problems when using the function programmatically (e.g. as part of a pipeline) when the length of the input is not known in advance (e.g. when the input vector is the result of some previous computation).

Since `getCARD2NCBIProtein()` is guaranteed to map each card id to a single protein id (many-to-one relationship), it should _always_ return a character vector of the same length as the input, no matter what that length is.

Now the one-to-many case is a different story. By default a function like `getNCBIProtein2CARD()` is still guaranteed to map each input id to at most one output id so there's no reason to not do the same as for `getCARD2NCBIProtein()`, that is, the function should _always_ return a character vector of the same length as the input, whatever that length is. It's _only_ when `exhaustiveMapping` is set to `TRUE` that the function should return a list. But again, this should not depend on the length of the input i.e. it should _always_ return a list, even when the input has length 1.

Hopefully this makes sense.

Best,
H.

**Comment by hpages** [comment]
Also it would be nice to show some examples that use `exhaustiveMapping=TRUE` in the man pages of the functions that support this argument. The examples should preferrably use an input vector that contains ids that illustrate the one-to-many situation, that is, the vector should contain a mix of ids that are mapped to 0, 1, or more than 1 id.

Thanks

**Comment by fersolesp** [comment]
Hi @hpages!

We understand your concern and have consequently added a parsing layer that takes care of homogenising outputs as suggested. Now we have:

```
interfaceFunction <- outputParser(vectorization(retryHandler(errorCacheHandling(memoise(originalFunction)))))
```

So, for simple mapping functions (single id mapped to a single id), like `getCARD2NCBIProtein()`, they now always return a character vector of the same length as the input. We have also substituted `NULL` values and `character(0)` (when no mapping is found) with `NA` in order to "fill the gaps" in return vectors and follow this widely consensual approach like other vectorized functions such as `sqrt()`, `toupper()` and `as.numeric()`. For example:

```
> getCARD2NCBIProtein(c("3003988", 'test' ,"3002535"))
[1] "APB03221.1" NA           "CAA38525.1"
```

The functions that may map one id to several ids have the same behaviour as the previous ones except if `exhaustiveMapping` is set to `TRUE`, when they return a list to avoid ids from mixing (always return a list even when input has length 1). Here are some examples with `getUniProt2KEGG()` function:

For id `Q6R7P5`, that has no translation:
```
> getUniProt2KEGG(c('Q6R7P5')) # Same as getUniProt2KEGG('Q6R7P5')
[1] NA

> getUniProt2KEGG(c('Q6R7P5'), detailedMapping = TRUE)
[1] NA

> getUniProt2KEGG(c('Q6R7P5'), exhaustiveMapping = TRUE)
[[1]]
[1] NA

> getUniProt2KEGG(c('Q6R7P5'), detailedMapping = TRUE, exhaustiveMapping = TRUE)
[[1]]
[1] NA
```

For `test` that fails because it is not a correct id:
```
> getUniProt2KEGG(c('test')) # Same as getUniProt2KEGG('test')
The given ID "test" is not registered in UniProt database
[1] NA

> getUniProt2KEGG(c('test'), detailedMapping = TRUE)
The given ID "test" is not registered in UniProt database
[1] NA

> getUniProt2KEGG(c('test'), exhaustiveMapping = TRUE)
The given ID "test" is not registered in UniProt database
[[1]]
[1] NA

> getUniProt2KEGG(c('test'), detailedMapping = TRUE, exhaustiveMapping = TRUE)
The given ID "test" is not registered in UniProt database
[[1]]
[1] NA
```

And for `A0A2R4PHC7` that has several possible translations:
```
> getUniProt2KEGG(c('A0A2R4PHC7')) # Same as getUniProt2KEGG('A0A2R4PHC7')
[1] "fcs:TRV642_2906"

# Note that it is a named vector
> getUniProt2KEGG(c('A0A2R4PHC7'), detailedMapping = TRUE)
$`0.5`
[1] "fcs:TRV642_2906"

> getUniProt2KEGG(c('A0A2R4PHC7'), exhaustiveMapping = TRUE)
[[1]]
[1] "fcs:TRV642_2906" "fcr:HYN56_14615" "fls:GLV81_10715" "ffl:HYN86_12595"

> getUniProt2KEGG(c('A0A2R4PHC7'), detailedMapping = TRUE, exhaustiveMapping = TRUE)
[[1]]
[[1]]$`0.5`
[1] "fcs:TRV642_2906" "fcr:HYN56_14615" "fls:GLV81_10715" "ffl:HYN86_12595"
```

When combining them in a single call:
```
> getUniProt2KEGG(c('Q6R7P5', 'test','A0A2R4PHC7'))
The given ID "test" is not registered in UniProt database
[1] NA                NA                "fcs:TRV642_2906"

# Note that it is a named vector
> getUniProt2KEGG(c('Q6R7P5', 'test','A0A2R4PHC7'), detailedMapping = TRUE)
The given ID "test" is not registered in UniProt database
                                                  0.5 
               NA                NA "fcs:TRV642_2906" 

> getUniProt2KEGG(c('Q6R7P5', 'test','A0A2R4PHC7'), exhaustiveMapping = TRUE)
The given ID "test" is not registered in UniProt database
[[1]]
[1] NA

[[2]]
[1] NA

[[3]]
[1] "fcs:TRV642_2906" "fcr:HYN56_14615" "fls:GLV81_10715" "ffl:HYN86_12595"

> getUniProt2KEGG(c('Q6R7P5', 'test','A0A2R4PHC7'), detailedMapping = TRUE, exhaustiveMapping = TRUE)
[[1]]
[1] NA

[[2]]
[1] NA

[[3]]
[[3]]$`0.5`
[1] "fcs:TRV642_2906" "fcr:HYN56_14615" "fls:GLV81_10715" "ffl:HYN86_12595"
```

We have already modified the whole test suite to work with the new output format and redone the man pages to cover it. We have also added some examples of `exhaustiveMapping` and `detailedMapping` in man pages but we have wrapped some of them in `\donttest` clauses (saw in Bioconductor instructions it is prefereable to `\dontrun`) because requesting `exhaustiveMapping = TRUE` means trying to retrieve all possible translations that in some cases may take a long time and would make `R CMD check` last too long . As stated before, despite of having some `\donttest` clauses, all functionality is tested once at some point (in unit or long tests). We hope you understand this decision.

Kind regards,

Fernando.

**Comment by hpages** [comment]
Thanks for the changes.

`getCARD2NCBIProtein(character(0))` and `getNCBIProtein2CARD(character(0))` should return `character(0)`, not `NA`. Again, this is following the model of what vectorized base functions like `sqrt()`, `toupper()`, and `as.numeric()` do. Note that this is completely standard behavior. Doing anything else will unnecessarily surprise the end-user and be error-prone when the function is used in the context of a complex pipeline.

`getNCBIProtein2CARD(character(0), exhaustiveMapping=TRUE)` should return an empty list (i.e. a list of length 0).

`getNCBIProtein2CARD(c("AAK64581", "test", "WP_010896559.1"), exhaustiveMapping=TRUE)` should return a list of length 3 (like it does), where the 2nd list element has length 0 (i.e. should be a `NULL` or `character(0)`). It should not be `NA`. The reason for this is that the length of the individual list elements should reflect the number of CARD ids associated with each protein id. In other words, the user should be able to quickly extract the number of CARD ids associated to each protein id by calling `lengths()` on the returned list. Right now, doing so returns the `c(1, 1, 1)` vector, but it should return `c(1, 0, 1)`.

Following the same logic, `getNCBIProtein2CARD("test", exhaustiveMapping=TRUE)` should return `list(NULL)` (or `list(character(0))`).

Also your code seems unnecessarily complex and is quite inefficient. For example, for `getNCBIProtein2CARD()`, all you need to do is something like this:
```
### We trust that 'ncbiId' is a single string and that 'exhaustiveMapping' is a single
### logical. No need to check that again.
.getNCBIProtein2CARD <- function(ncbiId, exhaustiveMapping=FALSE)
{   
    ncbiId <- gsub("\\..*$", "", ncbiId)
    if (ncbiId == "") {
        aroAccession <- character(0)
    } else {
        aroIndex <- .loadAROIndex()
        ok <- gsub("\\..*$", "", aroIndex$Protein.Accession) == ncbiId
        aroAccession <- aroIndex$ARO.Accession[ok]
    }
    if (exhaustiveMapping)
        return(aroAccession)
    if (length(aroAccession) == 0)
        return(NA_character_)
    aroAccession[[1]]
}

### Vectorized version of .getNCBIProtein2CARD().
getNCBIProtein2CARD <- function(ncbiId, exhaustiveMapping=FALSE)
{
    if (!is.character(ncbiId))
        stop("'ncbiId' must be a character vector")
    .checkBoolean(exhaustiveMapping, "exhaustiveMapping")
    .checkIfCARDIsDownloaded()
    if (exhaustiveMapping) 
        lapply(ncbiId, .getNCBIProtein2CARD, exhaustiveMapping=TRUE)
    else
        vapply(ncbiId, .getNCBIProtein2CARD, character(1),
               USE.NAMES=!is.null(names(ncbiId)))
}
```

This is about 9x faster than `ginmappeR:::getNCBIProtein2CARD()` when all the input is mapped, and hundreds times faster when the input contains invalid ids. The reason `ginmappeR:::getNCBIProtein2CARD()` is so slow when the input contains invalid ids is because it queries some online resource to check the id. Why would you do that if you already have a local copy of the CARD database?

All this is pretty standard business. As mentioned previously, there's no need to memoise `.getNCBIProtein2CARD()` or to use `Vectorize()`. As you can see vectorization can simply be done with `lapply()` or `vapply()`, which are very easy to use and give you total control.

It's important to keep your code as simple as possible (KISS principle).

We're getting closer, thanks!

**Comment by hpages** [comment]
Also it's a good idea to propagate the names on the input to the output (this is also standard behavior of vectorized base functions). The `getNCBIProtein2CARD()` function above does that:
```
ncbiId <- c(prot1="WP_010896559.1", prot2="test", prot3="AAG07988.1")

tolower(ncbiId)
#            prot1            prot2            prot3 
# "wp_010896559.1"           "test"     "aag07988.1" 


getNCBIProtein2CARD(ncbiId)
# Using a CARD database version downloaded on Tue 02/13/2024 17:18:24, please consider
# updating it with updateCARDDataBase() function.
#         prot1         prot2         prot3 
# "ARO:3004643"            NA "ARO:3004059" 

getNCBIProtein2CARD(ncbiId, exhaustiveMapping=TRUE)
# Using a CARD database version downloaded on Tue 02/13/2024 17:18:24, please consider
# updating it with updateCARDDataBase() function.
# $prot1
# [1] "ARO:3004643"
# 
# $prot2
# character(0)
# 
# $prot3
# [1] "ARO:3004059" "ARO:3004060"
```

**Comment by fersolesp** [comment]
Hi @hpages!

Thank you for the feedback. We have modified the behaviour of the parsing layer so it does follow all the standard ones suggested. For example, when providing `character(0)`, the functions now return `character(0)`. If the function has `exhaustiveMapping` set to `TRUE` it will return `list()`.

For ids with no translation it now returns `list(NULL)` when `exhaustiveMapping = TRUE`:

```
> getUniProt2KEGG(c('test'))
The given ID "test" is not registered in UniProt database
[1] NA

> getUniProt2KEGG(c('test'), exhaustiveMapping = TRUE)
The given ID "test" is not registered in UniProt database
[[1]]
[1] NULL
```

We have substituted `NA` in lists, too, for example:
```
> getNCBIProtein2CARD("test", exhaustiveMapping=TRUE)
[[1]]
NULL

> getNCBIProtein2CARD(c("AAK64581", "test", "WP_010896559.1"), exhaustiveMapping=TRUE)
[[1]]
[1] "ARO:3004568"

[[2]]
NULL

[[3]]
[1] "ARO:3004643"
```

With `character(0)` values in vectorized inputs, the functions follow the widely consensual approach of not returning anything for these values:

```
> getUniProt2KEGG(c('Q6R7P5', 'test', character(0), 'A0A2R4PHC7'), detailedMapping = TRUE, exhaustiveMapping = TRUE)
[[1]]
[1] NULL
 
[[2]]
[1] NULL
 
[[3]]
[[3]]$`0.5`
[1] "fcs:TRV642_2906" "fcr:HYN56_14615" "fls:GLV81_10715" "ffl:HYN86_12595"
```
Basically because it is technologically impossible as an input vector like `c('a', character(0), 'b')` is received like `c('a', 'b')` inside the function.

About `getNCBIProtein2CARD()` and specific functions that work with locally downloaded CARD database, it is true that the online check, in this case, of the NCBI id, is not necessary as we are using it to look into CARD. We thank you for pointing that out, we have deleted them. We have gone through the identifier checker functions and all of them are almost instantly executed except from NCBI checkers, because of its API's nature. Nevertheless, it is worth keeping them, because in case of invalid id, they prevent attempts to translate them, which takes even more time, and if they exist in the database, their verification is usually almost instantaneous.

Then, memoising the function and using `Vectorize()` may not be essential in this small set of functions that access CARD locally but we thought about package's homogeneity above all and having all functions the same interface and layered structure, specially the parsing and error handling layers. For practical purposes, having this features adds little or no further complexity.

Additionally, we have reviewed the whole codebase:
- `getCARD2NCBIGene()`: a coding error related to interfaces and original functions has been fixed and the effiency has also been improved by calling the auxiliar function `getNCBINucleotide2NCBIGene()` only when strictly needed.
- `.auxNCBI2CARD()`: this function that translates a bunch of NCBI ids to CARD was repeated in KEGG and UniProt files, so it has been moved to utils file, making the code clearer and simpler now.
- `getNCBIProtein2NCBINucleotide()` and `.getNCBINucleotide2NCBIProtein()`: fixed a small bug that under certain circunstances would not execute one of the mapping paths, losing the possibility of finding a translation if there is one.

On the other hand, the function `Vectorize()` is just a wrapper of `mapply()` but given specifically in R to vectorize functions so we see no necessity to change all code to use `vapply()` or `sapply()`, because it will be essentially the same. It would only be worth changing because of propagation of input names to the output, as suggested, but we consider it would be in conflict with parameter `detailedMapping`, that uses named vectors to indicate the origin of identifiers, and would make the user even more confused. In any case, if you consider it is worth adding, we would be glad to do it.

Thanks for your time,

Fernando.

**Comment by fersolesp** [comment]
> Hi @hpages!
> 
> Thank you for the feedback. We have modified the behaviour of the parsing layer so it does follow all the standard ones suggested. For example, when providing `character(0)`, the functions now return `character(0)`. If the function has `exhaustiveMapping` set to `TRUE` it will return `list()`.
> 
> For ids with no translation it now returns `list(NULL)` when `exhaustiveMapping = TRUE`:
> 
> ```
> > getUniProt2KEGG(c('test'))
> The given ID "test" is not registered in UniProt database
> [1] NA
> 
> > getUniProt2KEGG(c('test'), exhaustiveMapping = TRUE)
> The given ID "test" is not registered in UniProt database
> [[1]]
> [1] NULL
> ```
> 
> We have substituted `NA` in lists, too, for example:
> 
> ```
> > getNCBIProtein2CARD("test", exhaustiveMapping=TRUE)
> [[1]]
> NULL
> 
> > getNCBIProtein2CARD(c("AAK64581", "test", "WP_010896559.1"), exhaustiveMapping=TRUE)
> [[1]]
> [1] "ARO:3004568"
> 
> [[2]]
> NULL
> 
> [[3]]
> [1] "ARO:3004643"
> ```
> 
> With `character(0)` values in vectorized inputs, the functions follow the widely consensual approach of not returning anything for these values:
> 
> ```
> > getUniProt2KEGG(c('Q6R7P5', 'test', character(0), 'A0A2R4PHC7'), detailedMapping = TRUE, exhaustiveMapping = TRUE)
> [[1]]
> [1] NULL
>  
> [[2]]
> [1] NULL
>  
> [[3]]
> [[3]]$`0.5`
> [1] "fcs:TRV642_2906" "fcr:HYN56_14615" "fls:GLV81_10715" "ffl:HYN86_12595"
> ```
> 
> Basically because it is technologically impossible as an input vector like `c('a', character(0), 'b')` is received like `c('a', 'b')` inside the function.
> 
> About `getNCBIProtein2CARD()` and specific functions that work with locally downloaded CARD database, it is true that the online check, in this case, of the NCBI id, is not necessary as we are using it to look into CARD. We thank you for pointing that out, we have deleted them. We have gone through the identifier checker functions and all of them are almost instantly executed except from NCBI checkers, because of its API's nature. Nevertheless, it is worth keeping them, because in case of invalid id, they prevent attempts to translate them, which takes even more time, and if they exist in the database, their verification is usually almost instantaneous.
> 
> Then, memoising the function and using `Vectorize()` may not be essential in this small set of functions that access CARD locally but we thought about package's homogeneity above all and having all functions the same interface and layered structure, specially the parsing and error handling layers. For practical purposes, having this features adds little or no further complexity.
> 
> Additionally, we have reviewed the whole codebase:
> 
>     * `getCARD2NCBIGene()`: a coding error related to interfaces and original functions has been fixed and the effiency has also been improved by calling the auxiliar function `getNCBINucleotide2NCBIGene()` only when strictly needed.
> 
>     * `.auxNCBI2CARD()`: this function that translates a bunch of NCBI ids to CARD was repeated in KEGG and UniProt files, so it has been moved to utils file, making the code clearer and simpler now.
> 
>     * `getNCBIProtein2NCBINucleotide()` and `.getNCBINucleotide2NCBIProtein()`: fixed a small bug that under certain circunstances would not execute one of the mapping paths, losing the possibility of finding a translation if there is one.
> 
> 
> On the other hand, the function `Vectorize()` is just a wrapper of `mapply()` but given specifically in R to vectorize functions so we see no necessity to change all code to use `vapply()` or `sapply()`, because it will be essentially the same. It would only be worth changing because of propagation of input names to the output, as suggested, but we consider it would be in conflict with parameter `detailedMapping`, that uses named vectors to indicate the origin of identifiers, and would make the user even more confused. In any case, if you consider it is worth adding, we would be glad to do it.
> 
> Thanks for your time,
> 
> Fernando.

Hi @hpages!

In addition to last changes, we have just fixed a small error when downloading CARD database specifically in Windows systems.

Kind regards,

Fernando.

**Comment by hpages** [comment]
Hi Fernando,

1. Several `.R` files in the `R/` folder start with `source("R/utilsFunctions.R")`. This is considered bad practice. By default files in `R/` are collated in lexicographic order so just name your files according to the order that you wish to see (e.g. `00utils.R` instead of `utilsFunctions.R`). Alternatively, you can use the `Collate` field in `DESCRIPTION` if you don't want to rely on lexicographic order. See `1.1.1 The DESCRIPTION file` in the `Writing R Extensions` manual [here](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file) for more information.

2. Alway use `||` or `&&` in conditionals instead of `|` or `&`. The latters are vectorized versions of the former that will always evaluate the 2 sides of the operator, which is not what we want in a conditional.

3. Still a lot of inefficiencies and unnecessary obfuscation in your code e.g. why would internal helper `.getUniProt2KEGGDT()` collapse the character vector returned by `keggConv()` to immediately uncollapse it (with `strsplit()`) in case the vector was not empty:
    ```
    .getUniProt2KEGGDT <- function(upId){
        keggId <- paste(keggConv("genes", sprintf('uniprot:%s', upId)), collapse = ';')
        if(identical(keggId, character(0))|identical(keggId, "")){
            return(character(0))
        }else{
            return(c(strsplit(keggId,';',fixed = TRUE)[[1]]))
        }
    }
    ```
   What's wrong with simply doing:
    ```
    .getUniProt2KEGGDT <- function(upId){
        keggId <- keggConv("genes", sprintf('uniprot:%s', upId))
        if (length(keggId) == 0)
            return(character(0))
        keggId
    }
    ```
   Hopefully with time you will start to understand the benefits of writing code that is as simple as possible and you will fully embrace the KISS principle.

Please address 1. and 2. and the package will be good to go.

Thanks again,
H.

**Comment by hpages** [comment]
Or, since `keggConv()` is expected to always return a character vector, even when the length of the result is 0, the `if` statement is actually not needed:
```
.getUniProt2KEGGDT <- function(upId){
    keggConv("genes", sprintf('uniprot:%s', upId))
}
```
This is when one realizes that `getUniProt2KEGGDT()` is not doing much so its implementation should reflect that.

**Comment by fersolesp** [comment]
Hi @hpages!

Thank you for your suggestions, we have addressed them:

- We opted for naming the file `00utils.R` and deleted all `source("R/utilsFunctions.R")` imports.
- We replaced `|` and `&` with `||` and `&&`.

Thanks for your time,

Fernando.

**Comment by fersolesp** [comment]
Hi!

We have just fixed a small bug in getIdenticalProteins() function. It seems because of the package being '3a. accepted' I cannot push the changes to the Bioconductor git.

Kind regards,

Fernando.

**Comment by lshep** [comment]
It was pushed to git.bioconductor.org ,  it just meant you would not get a build on demand here.  I will be moving accepted packages into the daily builds later today or tomorrow; the change commit should be reflected there.

**Comment by fersolesp** [comment]
Oh okay! Thank you!!

**Comment by jwokaty** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/fersolesp.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("ginmappeR")`. The package 'landing page' will be created at

https://bioconductor.org/packages/ginmappeR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
